File: artifacts/product-reality/sprint-191/m02/verify-census.ts
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
const root = process.cwd();
const read = (file: string) => JSON.parse(fs.readFileSync(path.resolve(root, file), 'utf8'));
const before = read('artifacts/product-reality/sprint-190/m06/final/proof/component-census/report.json');
const after = read('artifacts/product-reality/sprint-191/m02/census/report.json');
const key = (row: any) => `${row.input.object}/${row.input.context}`;
const previous = new Map(before.allRows.map((row: any) => [key(row), row]));
const changes = [], singles = [];
assert.equal(after.totalSchemas, 77); assert.equal(after.greenTotalSchemas, 77);
assert.equal(after.totalCells, 154); assert.equal(after.greenTotalCells, 154);
for (const row of after.allRows) {
  const old: any = previous.get(key(row)); assert.ok(old);
  if (old.green !== row.green) changes.push({ schema: key(row), before: {green:old.green,errors:old.cells.map((cell:any)=>cell.errors)}, after: {green:row.green,errors:row.cells.map((cell:any)=>cell.errors)} });
  if (row.input.context === 'workflow') continue;
  for (const cell of row.cells) {
    const oldCell = old.cells.find((item: any) => item.framework === cell.framework);
    const oldArtifact = read(oldCell.response.path).artifact, newArtifact = read(cell.response.path).artifact;
    assert.equal(newArtifact.contentHash, oldArtifact.contentHash, `${key(row)}/${cell.framework} single-screen artifact must remain byte-identical`);
    singles.push({schema:key(row),framework:cell.framework,before:oldArtifact.contentHash,after:newArtifact.contentHash});
  }
}
assert.deepEqual(changes.map(row=>row.schema).sort(), ['Organization/workflow','User/workflow']);
assert.equal(singles.length,132);
fs.writeFileSync('artifacts/product-reality/sprint-191/m02/census-attribution.json',JSON.stringify({status:'passed',schemaCount:77,greenSchemas:77,generationCells:154,greenCells:154,singleScreenSchemas:66,unchangedSingleScreenCells:132,changes,singles},null,2)+'\n');
console.log('77/77 schemas; 154/154 cells; exactly two status changes; 132 single-screen artifact hashes unchanged.');

File: artifacts/product-reality/sprint-196/m07/verify-census.py
"""Bounded Sprint 195 census verifier: Sprint 196 changes rendering, not composition."""
import collections
import hashlib
import json
from pathlib import Path

base = Path('artifacts/product-reality/sprint-196/m07')
read = lambda file: json.loads(Path(file).read_text())
sha = lambda data: hashlib.sha256(data).hexdigest()
before = read('artifacts/product-reality/sprint-195/m07/component-census/report.json')
after = read(base / 'component-census/report.json')
classes = {row['id']: row['proposedClassification'] for row in read('packages/component-contracts/registry/component-reconciliation.proposed.v2.json')['rows']}

def nodes(node):
    return [node] + [child for item in node.get('children', []) for child in nodes(item)]

def normalized(schema):
    ids = {node['id']: f'node-{index}' for index, node in enumerate(node for screen in schema['screens'] for node in nodes(screen))}
    def clean(value):
        if isinstance(value, dict): return {key: clean(item) for key, item in value.items()}
        if isinstance(value, list): return [clean(item) for item in value]
        return ids.get(value, value) if isinstance(value, str) else value
    return json.dumps(clean(schema), sort_keys=True, separators=(',', ':')).encode()

assert after['greenTotalSchemas'] == 77 and after['greenTotalCells'] == 154
assert len(after['allRows']) == len(before['allRows']) == 77
key = lambda row: (row['input']['object'], row['input']['context'])
old_rows = {key(row): row for row in before['allRows']}
assert len(old_rows) == 77 and set(old_rows) == {key(row) for row in after['allRows']}
observed = []
for row in after['allRows']:
    old = old_rows[key(row)]
    for item in [old, row]:
        assert sha(Path(item['composition']['path']).read_bytes()) == item['composition']['sha256']
    left, right = read(old['composition']['path'])['schema'], read(row['composition']['path'])['schema']
    assert normalized(left) == normalized(right), f'Undeclared composed-schema movement: {key(row)}'
    assert {cell['framework'] for cell in row['cells']} == {'react', 'vue'}
    for cell in row['cells']:
        assert cell['status'] == 'ok' and cell['artifactPresent'] and not cell['errors']
        assert sha(Path(cell['response']['path']).read_bytes()) == cell['response']['sha256']
    counts = collections.Counter(node['component'] for screen in right['screens'] for node in nodes(screen))
    observed.append({'input': row['input'], 'before': old['composition'], 'after': row['composition'],
                     'beforeNormalizedHash': sha(normalized(left)), 'afterNormalizedHash': sha(normalized(right)),
                     'components': [{'component': name, 'count': count, 'proposedClassification': classes[name]} for name, count in sorted(counts.items())],
                     'class': 'unchanged-composed-schema', 'added': [], 'removed': []})
result = {'baselineHead': before['head'], 'head': after['head'], 'executionHead': after['head'],
          'changedSchemas': 0, 'unchangedSchemas': 77, 'generationCells': 154, 'passedGenerationCells': 154,
          'classes': {}, 'objectsGainingCharts': [], 'rows': [], 'allRows': observed, 'unattributedChanges': [],
          'normalization': 'Generated node IDs and exact references become traversal indices. All other composed-schema bytes participate. UTC generated artifact changes are separately bound by the m05 golden migration and m07 runtime proof.'}
(base / 'schema-movement.json').write_text(json.dumps(result, indent=2) + '\n')
original = read(base / 'saved-original/report.json'); successor = read(base / 'saved-successor/report.json')
hashes = lambda report: {row['schema'] + '.json': row['input']['sha256'] for row in report['rows']}
assert hashes(original) == hashes(read('artifacts/product-reality/sprint-186/m05/reachability/report.json'))
adoption = read('artifacts/product-reality/sprint-188/m01/adoption.json')
assert hashes(successor) == {name: value for name, value in adoption['after'].items() if name != '_index.json'}
assert original['reachable'] == 15 and successor['reachable'] == 16
live = Path('/Users/systemsystems/portfolio/Design-Tools/OODS-Forge/packages/mcp-server/.oods/schemas')
actual = {file.name: sha(file.read_bytes()) for file in sorted(live.glob('*.json'))}
assert actual == adoption['after'] and len(actual) == 17
(base / 'saved-compatibility.json').write_text(json.dumps({'sourceHead': after['head'], 'originalInputsUnchanged': True, 'successorInputsUnchanged': True,
    'liveStoreFiles': 17, 'changedLiveFiles': 0, 'liveStore': {'path': str(live), 'hashes': actual}, 'readOnly': True}, indent=2) + '\n')
print('77 unchanged composed schemas, 154 generation cells; original15/16, successor16/16;17 unchanged live store hashes.')
